playbooks/operationalizing/vault_integration.adoc

```diff
@@ -0,0 +1,169 @@
+
+== The credential management orchestration
+
+There are four actors in the orchestration that injects credentials in an OpenShift pod:
+ . The Vault instance
+ . The Vault Controller. The Vault controller is a trusted intermediary and has high privileges with Vault. In fact the Vault Controller can create new Vault tokens.
+ . An https://docs.openshift.com/container-platform/latest/architecture/core_concepts/containers_and_images.html[Init Container]. Init Containers are containers that are run before the main container of a pod is started. Their job is to set up the right environment for the main container. The design decision is to put the Vault orchestration logic in an Init Container so not to  have to add behaviour to the main container and therefore have greater reusability of existing images.
+ . The main container that needs the credentials.
+The below picture represents the orchestration:
+
+image::vault-controller.png[]
+
+ . An Init Container requests a https://www.vaultproject.io/docs/concepts/response-wrapping.html[wrapped token] from the Vault Controller over an encrypted connection. Wrapped credentials allow you to pass credentials around without any of the intermediaries having to actually see the credential.
+ . The Vault Controller retrieves the Pod details from the Kubernetes API server.
+ . If the Pod exists and contains the vaultproject.io/policies annotation, the Vault Controller calls Vault and generates a unique wrapped token with access to the https://www.vaultproject.io/docs/concepts/policies.html[Vault policies] mentioned in the annotation. This step requires trust on pod author to have used to right policies. The generated token has a configurable https://www.vaultproject.io/docs/concepts/lease.html[TTL].
+ . The Vault Controller "calls back" the Init Container using the Pod IP obtained from the Kubernetes API over an encrypted connection and delivers it the newly created wrapped token. Notice that the Vault Controller does not trust the pod, it only trusts the master API. 
+ . The Init Container unwraps the token to obtain a the Vault token that will allow to access the credentials.
+ . The Vault token is written to a well-known location in a volume shared between the two containers (https://kubernetes.io/docs/concepts/storage/volumes/#emptydir[emptyDir]) and the Init Container exits.
+ . The main container reads the token from the token file. Now the main container can use the token to retrieve all the secrets allowed by the policies considered when the token was created.
+ . If needed the main container renews the token to keep it from expiring.
+
+This integration mitigates some of the attack vector that we have seen before, in fact
+ . Credentials are encrypted at rest in Vault
+ . No openshift cluster admin can see the credentials, plus in vault you can create https://www.vaultproject.io/docs/concepts/seal.html[sharded master keys], so that no Vault admin can by themselves unencrypt the credentials.
```

Comment:
before
`so that no Vault admin can by themselves unencrypt the credentials.`
after
`so that no Vault admin can decrypt the secrets all by them self`